clippy_lints/src/single_field_patterns.rs

```diff
@@ -0,0 +1,264 @@
+            match pat.kind {
+                PatKind::Or(pats) => self.stack.extend(pats),
+                _ => return Some(pat),
+            }
+        }
+        unreachable!("Or always has 2 patterns, so one of the prior returns must return");
+    }
+}
+
+fn find_sf_lint<'hir>(
+    cx: &LateContext<'_>,
+    patterns: impl Iterator<Item = &'hir Pat<'hir>>,
+    leaf_sf: &impl Fn(&PatKind<'hir>) -> Option<SingleField>,
+) -> Option<(SingleField, Vec<(Span, String)>)> {
+    let fields = FlatPatterns::new(patterns).map(|p| {
+        (
+            p.span,
+            if matches!(p.kind, PatKind::Wild) {
+                Some(SingleField::Unused)
+            } else {
+                leaf_sf(&p.kind)
+            },
+        )
+    });
+    let mut spans = Vec::<(Span, String)>::new();
+    let mut the_one: Option<SingleField> = None;
+    for (target, sf) in fields {
+        if target.from_expansion() {
+            return None;
+        }
+        if let Some(sf) = sf {
+            match sf {
+                SingleField::Unused => {
+                    spans.push((target, String::from("_")));
+                },
+                SingleField::Id { pattern, .. } => {
+                    if let Some(str) = snippet_opt(cx, pattern) {
+                        spans.push((target, str));
+                    } else {
+                        return None;
+                    }
+                    if let Some(one) = the_one {
+                        if sf != one {
+                            return None;
+                        }
+                    } else {
+                        the_one = Some(sf);
+                    }
+                },
+            }
+        } else {
+            return None;
+        }
+    }
+    the_one.map(|one| (one, spans))
+}
+
+fn apply_lint_sf(cx: &LateContext<'_>, span: Span, sugg: impl IntoIterator<Item = (Span, String)>) {
+    span_lint_and_then(
+        cx,
+        SINGLE_FIELD_PATTERNS,
+        span,
+        "this single-variant pattern only matches one field",
+        |diag| {
+            multispan_sugg_with_applicability(diag, "try this", Applicability::MachineApplicable, sugg);
+        },
+    );
+}
+
+fn remove_deref<'a>(mut scrutinee: &'a Expr<'a>) -> &'a Expr<'a> {
+    // it would be wrong to convert something like `if let (x, _) = *a` into `if let x = *a.0`
+    while let ExprKind::Unary(UnOp::Deref, expr) = scrutinee.kind {
+        scrutinee = expr;
+    }
+    scrutinee
+}
+
+fn lint_sf<'hir>(
+    cx: &LateContext<'_>,
+    overall_span: Span,
+    scrutinee: &Expr<'_>,
+    patterns: impl Iterator<Item = &'hir Pat<'hir>>,
+) {
+    if scrutinee.span.from_expansion() {
+        return;
+    }
+    let scrutinee_name = if let Some(name) = snippet_opt(cx, remove_deref(scrutinee).span) {
+        name
+    } else {
+        return;
+    };
+    match cx.typeck_results().expr_ty(scrutinee).kind() {
+        ty::TyKind::Adt(def, ..) if def.is_struct() => {
```

Comment:
This should be re-exported.

```suggestion
        ty::Adt(def, ..) if def.is_struct() => {
```